gdal/ogr/ogrsf_frmts/flatgeobuf/geometryreader.cpp

```diff
@@ -259,7 +259,10 @@ OGRMultiPolygon *GeometryReader::readMultiPolygon()
     auto mp = std::unique_ptr<OGRMultiPolygon>(new OGRMultiPolygon());
     for (uoffset_t i = 0; i < parts->size(); i++) {
         GeometryReader reader { parts->Get(i), GeometryType::Polygon, m_hasZ, m_hasM };
-        auto p = std::unique_ptr<OGRPolygon>(reader.read()->toPolygon());
+        auto g = reader.read();
+        if (g == nullptr)
+            return nullptr;
+        auto p = std::unique_ptr<OGRPolygon>(g->toPolygon());
```

Comment:
you could probably just do mp->addGeometryDirectly(g->toPolygon()) . There's no way p can be nullptr if g has been tested for nullptr before
The toCurve() cases are different since they don't invoke GeometryReader with a given geometry type, hence they could get something different and need to do the dynamic_cast test. Here we know that we get either a polygon or null.

Reply:
Ok, reworked but it looked like it should use std::unique_ptr and release as in the other cases so that is what I did.